packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -529,20 +533,33 @@ const ComposedModalDialog = React.forwardRef<
   useEffect(() => {
     if (!open) return;
 
-    const handleEscapeKey = (event) => {
+    const handleEscapeKey = (event: globalThis.KeyboardEvent) => {
       if (match(event, keys.Escape)) {
-        event.preventDefault();
-        event.stopPropagation();
-        closeModal(event);
+        const allModalContainers = document.querySelectorAll(
+          '[role="dialog"][aria-modal="true"]'
+        );
+
+        // Proper visibility check that handles visibility:hidden
+        const visibleModals = Array.from(allModalContainers).filter((modal) => {
+          const styles = window.getComputedStyle(modal);
+          return styles.display !== 'none' && styles.visibility === 'visible';
+        });
```

Comment:
Could `enable-presence` help here? Or if we can remove the document event handler, I think this could be deleted?